```{r}
library(conflicted)
library(immunedeconv)
library(SingleR)
library(celldex)
library(readr)
library(dplyr)
library(tibble)
library(BiocParallel)
```

# Run CIBERSORT, xCell and singleR to obtain Macrophage scores

In this notebook, we run the deconvolution methods on the single-cell RNA-seq
data to obtain cell-type scores which can be used for cell-type annotation
in addition to the marker genes. 

```{r}
input_dir = "../../data/50_myeloid_cells/01_myeloid_subset/"
output_dir = "../../data/50_myeloid_cells/10_cell_type_scoring/"
register(MulticoreParam(workers=16), default=TRUE)
```

```{r}
set_cibersort_binary("../../lib/CIBERSORT/CIBERSORT.R")
set_cibersort_mat("../../lib/CIBERSORT/LM22.txt")
```

```{r}
X = read_csv(file.path(input_dir, "X.csv") %>% column_to_rownames("X1")
obs = read_csv(file.path(input_dir, "obs.csv") %>% column_to_rownames("X1")
```

## SingleR
```{r}
blueprint_encode = BlueprintEncodeData()
```

```{r}
res_encode = SingleR(X[,] %>% as.matrix, blueprint_encode, blueprint_encode$label.fine)
```

```{r}
res_df = res_encode$scores %>% as.data.frame()
res_df$labels = res_encode$labels
write_csv(res_df, file.path(output_dir, "singler_res.csv"))
```

## xCell
```{r}
expected_cell_types = c("Myeloid dendritic cell", "Myeloid dendritic cell activated", "Macrophage", "Macrophage M1", "Macrophage M2", "Monocyte", "Plasmacytoid dendritic cell")
xcell_res = deconvolute(X, method="xcell", expected_cell_types=expected_cell_types)
write_csv(xcell_res, file.path(output_dir, "xcell_res.csv"))
```

## CIBERSORT
```{r}
chunk_size = 150
chunks = seq(1, ncol(X), chunk_size)
cibersort_res = bplapply(chunks, function(i) {
  deconvolute(X[i:min(i+chunk_size, ncol(X))], method="cibersort") %>% column_to_rownames("cell_type")
}) %>% bind_cols()
write_csv(cibersort_res, file.path(output_dir, "cibersort_res.csv"))
```

## quanTIseq
```{r}
chunk_size = 150
chunks = seq(1, ncol(X), chunk_size)
quantiseq_res = bplapply(chunks, function(i) {
  deconvolute(X[i:min(i+chunk_size, ncol(X))], method="quantiseq") %>% column_to_rownames("cell_type")
}) %>% bind_cols()
write_csv(quantiseq_res, file.path(output_dir, "quantiseq_res.csv"))
```


